src/OpenTelemetry.Sampler.AWS/RulesCache.cs

```diff
@@ -0,0 +1,138 @@
+
+internal class RulesCache : IDisposable
+{
+    private const int CacheTTL = 60 * 60; // cache expires 1 hour after the refresh (in sec)
+
+    private readonly ReaderWriterLockSlim rwLock;
+
+    public RulesCache(Clock clock, string clientId, Resource resource, Trace.Sampler fallbackSampler)
+    {
+        this.rwLock = new ReaderWriterLockSlim();
+        this.Clock = clock;
+        this.ClientId = clientId;
+        this.Resource = resource;
+        this.FallbackSampler = fallbackSampler;
+        this.RuleAppliers = new List<SamplingRuleApplier>();
+        this.UpdatedAt = this.Clock.Now();
+    }
+
+    internal Clock Clock { get; set; }
+
+    internal string ClientId { get; set; }
+
+    internal Resource Resource { get; set; }
+
+    internal Trace.Sampler FallbackSampler { get; set; }
+
+    internal List<SamplingRuleApplier> RuleAppliers { get; set; }
+
+    internal DateTime UpdatedAt { get; set; }
+
+    public bool Expired()
+    {
+        this.rwLock.EnterReadLock();
+        try
+        {
+            return this.Clock.Now() > this.UpdatedAt.AddSeconds(CacheTTL);
+        }
+        finally
+        {
+            this.rwLock.ExitReadLock();
+        }
+    }
+
+    public void UpdateRules(List<SamplingRule> newRules)
+    {
+        // sort the new rules
+        newRules.Sort((x, y) => x.CompareTo(y));
+
+        List<SamplingRuleApplier> newRuleAppliers = new List<SamplingRuleApplier>();
+        foreach (var rule in newRules)
+        {
+            var currentStatistics = this.RuleAppliers
+                .FirstOrDefault(currentApplier => currentApplier.RuleName == rule.RuleName)
+                ?.Statistics ?? new Statistics();
+
+            var ruleApplier = new SamplingRuleApplier(this.ClientId, this.Clock, rule, currentStatistics);
+            newRuleAppliers.Add(ruleApplier);
+        }
+
+        this.rwLock.EnterWriteLock();
+        try
+        {
+            this.RuleAppliers = newRuleAppliers;
+            this.UpdatedAt = this.Clock.Now();
+        }
+        finally
+        {
+            this.rwLock.ExitWriteLock();
+        }
+    }
+
+    public SamplingResult ShouldSample(in SamplingParameters samplingParameters)
+    {
+        foreach (var ruleApplier in this.RuleAppliers)
+        {
+            if (ruleApplier.Matches(samplingParameters, this.Resource))
+            {
+                return ruleApplier.ShouldSample(in samplingParameters);
+            }
+        }
+
+        // ideally the default rule should have matched.
+        // if we are here then likely due to a bug.
+        AWSSamplerEventSource.Log.InfoUsingFallbackSampler();
+        return this.FallbackSampler.ShouldSample(in samplingParameters);
+    }
+
+    public void Dispose()
+    {
+        this.Dispose(true);
+        GC.SuppressFinalize(this);
+    }
+
+    internal DateTime GetUpdatedAt()
+    {
+        this.rwLock.EnterReadLock();
+        try
+        {
+            return this.UpdatedAt;
```

Comment:
Is locking around returning a variable needed?

Also, I don't think GetUpdatedAt is used.